blog/2017-12-07-streaming-to-elasticsearch.adoc

```diff
@@ -0,0 +1,248 @@
+
+* https://hub.docker.com/r/debezium/zookeeper/[Apache ZooKeeper]
+* https://hub.docker.com/r/debezium/kafka/[Apache Kafka]
+* An https://github.com/debezium/debezium-examples/tree/master/unwrap-smt/debezium-jdbc[enriched] Kafka Connect / Debezium https://hub.docker.com/r/debezium/connect/[image] with changes
+** PostgreSQL JDBC driver placed into `/kafka/libs` directory
+** https://docs.confluent.io/current/connect/connect-jdbc/docs/index.html[Kafka Connect JDBC Connector] (developed by https://www.confluent.io/[Confluent]) placed into `/kafka/connect/kafka-connect-jdbc` directory
+* Pre-populated MySQL used in our link:docs/tutorial[tutorial]
+* Empty PostgreSQL
+* Empty ElasticSearch
+
+We again use the `UnwrapFromEnvelope` transformation to automatically extract a new row record and thus effectively convert the Debezium complex record into a simple one consumable by other connectors. Both JDBC and Elasticsearch connectors expect the same format.
+
+== Example
+Let's move directly to our exmaple as that's where the changes are visible.
+
+First of all we need to deploy all components.
+[source,bash,indent=0]
+----
+export DEBEZIUM_VERSION=0.6
+docker-compose up
+----
+
+When all components are started we are going to register the Elasticsearch Sink connector writing into the ElasticSearch instance.
+We want to use the same key (primary id) in the source and both PostgreSQL and Elasticsearch.
+[source,bash,indent=0]
+----
+curl -i -X POST -H "Accept:application/json" -H  "Content-Type:application/json" http://localhost:8083/connectors/ -d @es-sink.json
+----
+
+Using this registration request:
+
+[source,json,indent=0]
+----
+{
+    {
+        "name": "elastic-sink",
+        "config": {
+            "connector.class": "io.confluent.connect.elasticsearch.ElasticsearchSinkConnector",
+            "tasks.max": "1",
+            "topics": "customers",
+            "connection.url": "http://elastic:9200",
+            "transforms": "unwrap,key",
+            "transforms.unwrap.type": "io.debezium.transforms.UnwrapFromEnvelope",           (1)
+            "transforms.key.type": "org.apache.kafka.connect.transforms.ExtractField$Key",   (2)
+            "transforms.key.field": "id",                                                    (2)
+            "key.ignore": "false",                                                           (3)
+            "type.name": "kafka-connect"                                                     (4)
+        }
+    }
+}
+----
+
+The request configures these options:
+
+1. unwrapping Debezium's complex format into a simple one
+2. extracting `id` field from the key `struct`, then the same key is used for the source and both destinations.
+This is to overcome limitation of Elasticsearch connector that supports only `int`s and `string`s as keys.
+3. use key from the event instead of generating a synthetic one
+4. type under which the events will be registered in Elasticsearch
```

Comment:
Wouldn't it make more sense to use something like "customer"?

Reply:
Yes and no. I kept it intentionally generic. It make sens to set it to customer if you are lsitening only on customers topic. If you add orders topic then th reasoning breaks. I'd say that type here describes it is a change data and the actual type (the originating  table) is defined by index name.

Reply:
How is the index name derived btw.? Is it based on the connector name? The reason I'm asking is that you can't mix multiple types in a single ES index in current versions. So if the index name is derived from the connector instance name, it'd mean that one would typically have to use one connector per table/topic to be routed to ES.

Reply:
If `topic.index.map` is not set then the name of tipic is used, which is as good as it can be for this case.

Reply:
Ok, got it. The type name is obsolete essentially as of ES 6, but I reckon the sink connector currently still requires it.

> I'd say that type here describes it is a change data and the actual type (the originating table) is defined by index name.

That's not the meaning of "type" in the ES context. It really refers to one type of documents within an index, so you could e.g. have documents of type "b2b_customer" and "b2c_customer" within one index. The type mapping associated with such type describes the fields of documents of that type. Mixing multiple types within one index won't be supported going forward.

Bottom line, using "customer" is what makes most sense to me for the example.